std/format.d

```diff
@@ -4362,14 +4362,44 @@ T unformatValue(T, Range, Char)(ref Range input, ref FormatSpec!Char spec)
     return parse!T(input);
 }
 
+private T rawRead(T, Range)(ref Range input)
+    if (isSomeString!Range || ElementType!(Range).sizeof == 1)
+{
+    union X
+    {
+        ubyte[T.sizeof] raw;
+        T typed;
+    }
+    X x;
+    foreach (i; 0 .. T.sizeof)
+    {
+        static if (isSomeString!Range)
+        {
+            x.raw[i] = input[0];
+            input = input[1 .. $];
+        }
+        else
+        {
+            // TODO: recheck this
+            x.raw[i] = cast(ubyte) input.front;
+            input.popFront();
+        }
+    }
+    return x.typed;
+}
+
 /**
    Reads an integral value and returns it.
  */
 T unformatValue(T, Range, Char)(ref Range input, ref FormatSpec!Char spec)
     if (isInputRange!Range && isIntegral!T && !is(T == enum))
```

Comment:
Consider this code:

```d
void main()
{
	import std.format;
	double[] array;
	FormatSpec!char sp;
	auto x = unformatValue!int(array, sp);
}
```

This fails to compile (try it!), and the error messages point inside the standard library code. We never want to allow that to happen. Instead, we want to point at the user code "You passed the wrong arguments". That's why we need all templates in the stdlib have constraints that describe exactly the inputs they support.

There are clearly still places where existing code does not satisfy this, but at least we should make sure new code has the appropriate constraints. You may want to fix the problem for the other overloads as well!